[WIP] Enable importing single dossiers with votes

njoyard opened the pull request:
*Work in progress, do not merge*

This PR adds a new script to import a single dossier with votes (from http://parltrack.euwiki.org/dossier/YYYY/NNNN(XXX)?format=json URLs). The existing import code has been moved to a common lib module in order to enable usage from any script.

Tests have also been moved to django.test.TestCase. There are a bunch of warnings but they come from code in django-representatives.

The thing works, but uses exactly the same import code as previously existing scripts. I will try to rewrite those to be less DB-intensive soon.

njoyard commented:
Note that I should have made 2 PRs (one for the single import, one for moving to django TestCases). But as the single import update adds a test case, that would have implied writing it twice. I can still separate those if you prefer.

jpic commented:
I don't know why we lost pytest support here, since it wouldn't prevent using django.test.TestCase, but it doesn't matter ^^

If I really had to think of something to say about this PR I'd say you could squash  e9dc070 into 76c8a9c since they are both about migrating from pytest to django test runner, and 8973e6d into 55d804a since they are both about adding the new entry point - the refactor alone is not really a feature, but i'm feature-commit oriented so don't take my word for it, there are many ways to use git and i don't think it's so important. If you want to try, maybe with `git rebase -i origin/master` you could mark e9dc070 and 8973e6d as fixups and mark 55d804a for rewording to describe it like "add entry point for single dossier import". I don't see the code that would download the data from parltrack, maybe you plan on making a PR for this afterwards ?

njoyard commented:
The move away from pytest is just to simplify things, as we already have django-admin test anyway. That's not mandatory, but I figured pytest didn't add anything here anyway.

I plan to add data download later, but I'm not sure where it's best fitted. The existing data update triggers (ie. the actual download from parltrack and launching the parltrack_import_* scripts) are in political_memory, shouldn't we do the same for this new feature (in which case this PR needs to be merged first)? Or should we instead move all download scripts to this repo (and django-representatives)?

jpic commented:
Well, I thought we wanted to be able to trigger a single proposal votes synchronisation uppon Recommendation update in memopol. So in memopol it would be something like:

political_memory/votes/models.py:

```
def sync_votes(sender, instance=None, **kwargs):
    # instance is a Recommendation here, because this callback is a receiver for sender=Recommendatino
    instance.proposal.sync_votes()
    # also update scores here ^^
post_save.connect(sync_votes, sender=Recommendation)
```

But the logic to download the votes from parltrack into the proposal would be completely independant from memopol here wouldn't it ?

jpic commented:
Yeah funny thing about pytest is that i used to always use the plain old test runner from django but in open source projects at some point some contributor arrives and says how much better is pytest and i'm like "yeah well ok i know i wanted to use pytest but i thought contributors would be more comfortable with nose-ish test runner so let's move to pytest then" and this time the opposite happened, it's like i never seem to choose the right one on the first try hahahahaha

njoyard commented:
Yeah, that would be a python download, not a shell + unxz download as is done in memopol now.

The problem is, score update logic is in memopol, not in django-representatives-votes. So we have two options:

* download the json from parltrack in memopol, pass it to django-representatives-votes to update votes, then update the scores in memopol
* call an update method in django-representative-votes from memopol, download the json from django-representative-votes, then update the scores in memopol

The choice is not very critical here, as in the end the only thing that changes is "who knows the parltrack URL between memopol and django-representatives-votes".

Then, the other question to ask is (and that's maybe due to my lack of knowledge of the data model, especially the exact role of each entity): when an admin updates a recommendation, do we want to update votes for the whole dossier (as we'll have to download the full dossier json anyway) or only the proposal related to the recommendation (and btw, votes are related to proposals, not dossiers, right?) ?

About the test runner, I've not touched any line of python for maybe 5 years before coming here to help you guys, so I'm probably not the best one to trust here. I just think, at least for this repository, pytest adds exactly nothing. But maybe you have future plans that only pytest can achieve, or maybe django-admin tests is sometimes buggy and that's why people use pytest instead, I have no idea.

Finally, I'd just like to mention that I mainly do PRs to give you guys a chance to see what I'm in the process of doing, and confirm I'm going in the right direction, before getting too carried away with 10s of changed files (again, my python is rusty, my django is brand new here, so I need you to confirm that I'm not doing crap :P). I prefer to do small PRs often, even if they're not getting merged, to get your feeling on things while the cost of redoing everything is still small :)

jpic commented:
I see that you want to download the json in memopol to be consistent with the other imports. In fact, we used to download the json in django-representative{,-votes} but migrated to a piped unxz solution to save disk space on OpenShift because we only have 1gig per free account. Since we didn't maintain the code to download and unxz from django-representatives{,-votes}, it was removed.

In general, we should try to have as few code as possible in Django projects, because they glue apps and are harder to test. We should try to encapsulate as much as possible in apps, because they have isolated dependencies and a clear scope which makes it easier to test.

If memopol could use a "single proposal sync" function, it's not specific to memopol, other projects could use it too -> that's why it makes more sense to encapsulate this feature inside django-representatives{,-votes}, even though I deleted the download code for the full imports because not-tested / no-maintainer.

Small PRs are good ;) At my day job we each have at least one big PR, but it's not for merging -> we extract feature by feature from our big PR into new smaller PR and once we've merged every feature from small PRs we close the big PR, we're pretty happy with that model so far so maybe you'd like to try that too, just FYI ^^

njoyard commented:
Okay, so I'll add parltrack sync here :)

njoyard commented:
Tests fail because parltrack is down. Is this what you expected ? (see changes to models.py)

Todo:
* Add single proposition sync
* Get rid of parltrack.euwiki.org dependency for tests (we'll need some kind of pluggable JSON data source for that), because we don't want tests to fail because 1. parltrack is down or 2. parltrack data changed and makes our test expectations wrong
* Small refactor to avoid method-level import in models.py (was done to resolve circular dependency [models -> sync -> models] temporarily without having flake8 complaining)
* Rebase on master and reorganize commits

jpic commented on .gitignore:
```diff
@@ -1,3 +1,4 @@
+test.db
```
Shouldn't tests be running in a :memory: database ?

njoyard replied:
They probably should, but they don't. Including when using py.test.

jpic commented on .travis.yml:
```diff
@@ -1,20 +1,18 @@
 language: python
 env:
-- DJANGO="django>1.8,<1.9" DJANGO_SETTINGS_MODULE=representatives_votes.tests.settings
+- DJANGO="django>1.8,<1.9" DJANGO_SETTINGS_MODULE=test.settings
```
Why aren't we bundling the tests in the package anymore ?

njoyard replied:
From the examples I've seen it seems good practice to keep tests in a separate app ; that way, running `django-admin test` on a dependant app won't run our tests.

jpic commented on .travis.yml:
```diff
@@ -1,20 +1,18 @@
 - flake8 representatives_votes/ --exclude migrations --ignore E128
-- py.test
-- cat representatives_votes/contrib/parltrack/tests/dossiers_fixture.json | parltrack_import_dossiers
-- cat representatives_votes/contrib/parltrack/tests/votes_fixture.json | parltrack_import_votes
```
I don't understand, I thought it was cool to keep parltrack optional, and loosely coupled from the app's core, so that we could use other sources (ie. for french parliament data).

njoyard replied:
Hmmm I don't quite see why you ask that question, but this PR does not really change the way parltrack and the app are coupled, I even tried to decouple it more (by adding JSON data sources).

Those lines you're commenting about are just getting rid of unneeded tests (those are run by `django-admin test` now).

jpic commented on representatives_votes/contrib/parltrack/import_votes.py:
```diff
@@ -1,185 +1,23 @@
-        if vote_data.get('ep_id', None) is None:
-            return
-        return self.cache['meps'].get(int(vote_data['ep_id']), None)
+vote_pre_import = vote_signal
```
Why ?

When does "vote_signal" happen ? Before or after it's imported ? :D

njoyard replied:
This signal existed before, just with a different name.

`vote_signal` happens when the previously existing `vote_pre_import` signal happened; If I get it right, it runs after updating vote counts in a proposal and before updating votes. It also enables receivers to cancel the update.

Since I moved actual updating code from import_*.py scripts to a common lib.py code, I had to move the signal as well. But this `vote_pre_import` signal was exported and used outside, so that's what I had to do to keep it the same way.

jpic commented on representatives_votes/contrib/parltrack/tests/test_import.py:
```diff
@@ -1,52 +0,0 @@
-    call_command('loaddata', os.path.join(os.path.dirname(__file__),
-        'dossiers_expected.json'))
-
-    _test_import('votes', import_votes.main)
```
Why is this gone ?

njoyard replied:
This is now in the 'test' app.

jpic commented:
Hehe, yeah the network is the number 1 enemy to the tester !!

Initially, the parltrack import test was done because we had no other way to cover the code, but now that tests were added with fixtures we don't need it anymore !

I think someone forgot to tell you that it was a requirement to keep parltrack specific code separated from the app's core, so that we could add other import scripts from other sources (ie. french parliament).

I completely fail to understand why the tests were removed from the python package, does that mean that we're unable to run tests if we just installed the package with pip ?

I also fail to understand why you were unhappy with py.test and decided to remove it in this PR which was initially about just adding a couple of functions.

What's the use case for the import_single_dossier.py command ? I thought you just wanted to add a function to sync a single dossier, and that we'd figure out later how we'd call it - we thought we'd use a separate worker at some point, and now i wonder if having a cron in memopol that would just call it for every proposals that have a recommendation wouldn't be just what we need.

I'd like to give it a try myself and then maybe we could compare both PRs and see what's best, what do you think ?

jpic commented on test/tests.py:
```diff
@@ -0,0 +1,89 @@
+        dossier.refresh_from_db()
+
+        exp = "Agenda for change: the future of EU development policy"
+        assert dossier.title == exp
```
If you're not using py.test, then assertEquals() and other xUnit-family assertion commands.

For example:

```
import unittest


class SomeTest(unittest.TestCase):
    a = True
    b = False

    def test_assert(self):
        assert self.a == self.b

    def test_xunit_assert(self):
        self.assertEqual(self.a, self.b)
```

With the unittest module, we just get an AssertionError which is not descriptive test_assert:

```
$ python -m unittest test
FF
======================================================================
FAIL: test_assert (test.SomeTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "test.py", line 9, in test_assert
    assert self.a == self.b
AssertionError

======================================================================
FAIL: test_xunit_assert (test.SomeTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "test.py", line 12, in test_xunit_assert
    self.assertEqual(self.a, self.b)
AssertionError: True != False

----------------------------------------------------------------------
Ran 2 tests in 0.000s

FAILED (failures=2)
```

With py.test, we get descriptive output for both:

```
[env] 03/01 2016 21:17:44 jpic@lue ~ 
$ py.test test.py
==================================== test session starts ====================================
platform linux2 -- Python 2.7.11, pytest-2.8.2, py-1.4.30, pluggy-0.3.1
rootdir: /home/jpic, inifile: 
plugins: django-2.8.0
collected 2 items 

test.py FF

========================================= FAILURES ==========================================
___________________________________ SomeTest.test_assert ____________________________________

self = <test.SomeTest testMethod=test_assert>

    def test_assert(self):
>       assert self.a == self.b
E       AssertionError: assert True == <test.SomeTest testMethod=test_assert>.b
E        +  where True = <test.SomeTest testMethod=test_assert>.a

test.py:9: AssertionError
________________________________ SomeTest.test_xunit_assert _________________________________

self = <test.SomeTest testMethod=test_xunit_assert>

    def test_xunit_assert(self):
>       self.assertEqual(self.a, self.b)
E       AssertionError: True != False

test.py:12: AssertionError
================================= 2 failed in 0.02 seconds ==================================
```

I think that's one of the reasons py.test is hype nowadays ^^

jpic replied:
unittest

![2016-01-03-212433_621x385_scrot](https://cloud.githubusercontent.com/assets/94636/12080626/7c23971a-b260-11e5-9027-8c4c38d7e1d3.png)

pytest:

![2016-01-03-212442_873x538_scrot](https://cloud.githubusercontent.com/assets/94636/12080629/8d3c3f8e-b260-11e5-861d-d5436f75af49.png)

What's best ? no color + self.assert* everywhere or pytest ?

Note that you can also use unittest-style classes with pytest if that's what you're after !

njoyard replied:
That's a good point, I'll get py.test back :P

jpic commented:
At work, when we get to the point we have such a big PR, we mark it as "WIP/NOMERGE" and do another set-of PR with just one small piece of it, so that we can merge small PRs one by one, and once everything has been isolated in small PRs we close the big one, does that sound good to you ?

njoyard commented:
> Initially, the parltrack import test was done because we had no other way to cover the code, but now that tests were added with fixtures we don't need it anymore !
> 
> I think someone forgot to tell you that it was a requirement to keep parltrack specific code separated from the app's core, so that we could add other import scripts from other sources (ie. french parliament).

I understood that, and I tried to keep it as decoupled as it was before. Is there somewhere you see that I coupled it more than it already was?

> I completely fail to understand why the tests were removed from the python package, does that mean that we're unable to run tests if we just installed the package with pip ?

I addressed that in a comment above already, but long story short, it seems good practice to keep tests out of what we "publish" to the outside of the module.  I *think* what I've done will prevent testing the module when installed with pip (not sure, have not tried it) but again, it seems a good idea to me.  We should not make apps that use our module fail their tests just because ours fail.

> I also fail to understand why you were unhappy with py.test and decided to remove it in this PR which was initially about just adding a couple of functions.

Initially I had not understood you could use django TestCases with it.  Plus I didn't see py.test adding any value from simply using django-admin test.  So I moved the tests away from py.test, and it sticked with the PR  :)

But as you showed in commit comments, py.test indeed has added value over django-admin test.  I'll get it back soon.

> What's the use case for the import_single_dossier.py command ? I thought you just wanted to add a function to sync a single dossier, and that we'd figure out later how we'd call it - we thought we'd use a separate worker at some point, and now i wonder if having a cron in memopol that would just call it for every proposals that have a recommendation wouldn't be just what we need.

This command is not there anymore.  It's just an import_* module as the others, with a `main(Stream)` entry point.  It mainly keeps code homogenous and allows testing all import/update procedures using the same _test_import helper (that was there before).

> I'd like to give it a try myself and then maybe we could compare both PRs and see what's best, what do you think ?

I won't get offended if you do things better than me :)

> At work, when we get to the point we have such a big PR, we mark it as "WIP/NOMERGE" and do another set-of PR with just one small piece of it, so that we can merge small PRs one by one, and once everything has been isolated in small PRs we close the big one, does that sound good to you ?

Agreed, marked the PR as WIP.

Cheers,

jpic commented:
> I understood that, and I tried to keep it as decoupled as it was before. Is there somewhere you see that I coupled it more than it already was?

Well the main test module as far as i understand (test/tests.py) includes parltrack-specific tests, which means, and when we bundle the [DRF code](https://github.com/political-memory/toutatis/tree/master/api), we'll have:

- move tests/test.py tests/tests.py for representatives.contrib.parltrack
- tests/api.py for representatives.api

Maybe we could have it in tests/contrib/test_parltrack.py then ?

> I addressed that in a comment above already, but long story short, it seems good practice to keep tests out of what we "publish" to the outside of the module. I think what I've done will prevent testing the module when installed with pip (not sure, have not tried it) but again, it seems a good idea to me. 

I did not find the comment, I think it might be right to not publish the test code since that's the way in most repos even though I find it a bit a shame. I think we should take the time to discuss this, so that we don't keep going back and forth on this :D For the record, I use to do this because in django-autocomplete-light, I initially did not include the ability to run tests from outside the repo, but a user (a django core dev ^^) complained about it, I think it's because they wanted to know if their environment broke the app, and the issue is still open, I lament not having done this in the beginning ! Now we have like 500 tests and it's to hard to change that without rewriting most of the test code (which has partly been done with that in mind from v1 to v2, but its still not sufficient).

 So if I want to the the module i have installed with pip, it would look like this now ?

- pip install the module,
- clone the repository,
- checkout the version of the module in the repository,
- go in the tests directory of the repo,
- hope that I have not installed a package with the name of `settings`,
- run `DJANGO_SETTINGS_MODULE=settings django-admin test representatives_votes` ?

With test publishing, we can:

- pip install the module,
- run `DJANGO_SETTINGS_MODULE=representatives.tests.settings py.test ~/path/to/module/tests` ?

Another reason I did that is that it seemed to be a [standard django to have tests inside app.tests (https://docs.djangoproject.com/en/1.9/intro/tutorial05/#create-a-test-to-expose-the-bug). We all know that Django is not right on everything, however we tend to stick to its practices to make it easy for other django users.

As a side but related note I find that python modules usually have docs in a separated directory, and that often I visit their code in site-packages and wish docs were installed there too, ie. i'm working offline in a train or plane or in the forest (lol), because i need internet to find and clone the repo, checkout the right version to have the docs offline, for next time. Did this ever happen to you ? I find it a bit of a shame, it's like when some distro installed a command but without its documentation, I can see how this makes sense on embedded systems with low resources - although compiling with -Os, stripping debug symbols, etc, etc would make sense to do in this case - but not really in other environments. So, we all do this "not publishing docs with python packages" but is it really practical ?

> We should not make apps that use our module fail their tests just because ours fail.

Could you share an example for this because I don't understand how that's possible.

We should certainly not publish a version with failing tests, **ever**, how could a user know if his environment broke the tests if we should code that doesn't pass them in the first place ?

Overall, I'm really confused about this topic now as you might have figure LOL
I could live with this change though, shouldn't we change this in django-representatives-votes and political-memory too, to keep the repos consistent with each other ? Would it be ok to postpone that and maybe ask on stackoverflow and / or mailing list to see what other folks think about this before taking a decision ? Sorry but I'd really like to make sure that if we change that, we're not going to change it again :D

> I won't get offended if you do things better than me :)

That's already better than me LOL

Your code is good, I really like the test code ! I don't think it's necessary that I duplicate the effort in another PR after all because I don't think I could do any better, I was just thinking about showing that this could be split into smaller PRs but you said you liked that idea too, it just means that one of the PR will have to be rebased but that's not a big deal I think.

njoyard commented:
I'm not trying to force my choices upon you, just explaining why I did things this way. Indeed, such changes need a bit of thought before committing fully into them.

I'm closing this PR for now but keep the branch for future reference. I have a new PR ready that's more straight to the point, I think.

> in the forest

That's all I'm going to remember from this discussion now :D
